src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -515,9 +522,27 @@ public bool Matches(TableEntry entry, IEqualityComparer<T> equalityComparer)
 
             public OneOrMany<T> Items => _items;
 
-            public TableEntry AsCached() => new(_items, s_allCachedEntries);
+            public TableEntry AsCached()
+            {
+                if (!_anyRemoved)
+                {
+                    return new TableEntry(_items, s_allCachedEntries, anyRemoved: false);
+                }
+
+                var itemBuilder = ArrayBuilder<T>.GetInstance();
+                for (int i = 0; i < this.Count; i++)
+                {
+                    if (this.GetState(i) != EntryState.Removed)
+                    {
+                        itemBuilder.Add(this.GetItem(i));
+                    }
+                }
+
+                Debug.Assert(itemBuilder.Count < this.Count);
+                return new TableEntry(OneOrMany.Create(itemBuilder.ToImmutableArray()), s_allCachedEntries, anyRemoved: false);
+            }
 
-            public TableEntry AsRemoved() => new(_items, s_allRemovedEntries);
+            public TableEntry AsRemovedDueToInputRemoval() => new(_items, s_allRemovedDueToInputRemoval, anyRemoved: true);
```

Comment:
Not related to this PR (ie. don't fix, but keep in mind in future changes): we have a policy to avoid target-typed `new` in the compiler codebase, consider fixing the few that snuck in when you encounter them and touch that code.